marimo/_runtime/dataflow/graph.py

```diff
@@ -5,47 +5,79 @@
 
 import threading
 from dataclasses import dataclass, field
-from typing import TYPE_CHECKING, Callable, Literal, Optional
+from typing import TYPE_CHECKING, Callable, Literal, Optional, cast
 
 from marimo import _loggers
 from marimo._ast.compiler import code_key
+from marimo._ast.sql_visitor import SQLTypes
 from marimo._ast.variables import is_mangled_local
-from marimo._runtime.dataflow import edges
-from marimo._runtime.dataflow.cycles import CycleTracker
+from marimo._runtime.dataflow.dag import MutableDirectedGraph
 from marimo._runtime.dataflow.definitions import DefinitionRegistry
-from marimo._runtime.dataflow.topology import (
-    GraphTopology,
-    MutableGraphTopology,
-)
+from marimo._runtime.dataflow.topology import GraphTopology
 
 if TYPE_CHECKING:
     from collections.abc import Mapping
 
     from marimo._ast.cell import CellImpl
+    from marimo._ast.sql_visitor import SQLRef
     from marimo._ast.visitor import ImportData, Name, VariableData
     from marimo._types.ids import CellId_t
 
 LOGGER = _loggers.marimo_logger()
 
 
+def _resolve_variable_name(
+    name: Name,
+    other_cell: CellImpl,
+    sql_ref: Optional[SQLRef],
+    sql_matches: list[tuple[set[CellId_t], Name]],
+) -> Name:
+    """
+    Resolve the variable name to use when checking if it exists in another cell.
+
+    For regular (non-SQL) references, returns the original name unchanged.
+    For SQL hierarchical references, finds the variable name from sql_matches that
+    is actually defined in the other_cell.
+
+    Example:
+        cell_1: CREATE SCHEMA schema_name
+        cell_2: CREATE TABLE schema_name.table_name
+        cell_3: FROM schema_name.table_name SELECT *
+
+        When cell_3 references "schema_name.table_name":
+        - For cell_1: returns "schema_name" (what cell_1 actually defines)
+        - For cell_2: returns "table_name" (what cell_2 actually defines)
+    """
+    if not sql_ref or name in other_cell.variable_data:
+        return name
+
+    # For SQL hierarchical references, find the matching variable name
+    for _, matching_variable_name in sql_matches:
+        if matching_variable_name in other_cell.variable_data:
+            return matching_variable_name
+
+    return name
+
+
 @dataclass(frozen=True)
-class DirectedGraph(GraphTopology):
```

Comment:
Just noting that we import `DirectedGraph` in marimo-lsp. Should be ok since we pin marimo upper bound in our rolling releases now, but wanted to flag.